plugins/thunderbolt/fu-thunderbolt-image.c

```diff
@@ -0,0 +1,571 @@
+	 GError                        **error)
+{
+	g_autoptr(GByteArray) controller_data = NULL;
+	g_autoptr(GByteArray) image_data      = NULL;
+
+	controller_data = read_location (location, controller_fw, error);
+	if (controller_data == NULL)
+		return FALSE;
+
+	image_data = read_location (location, image_fw, error);
+	if (image_data == NULL)
+		return FALSE;
+
+	*result = memcmp (controller_data->data, image_data->data, location->len) == 0;
+	return TRUE;
+}
+
+static void
+read_bool (const FuThunderboltFwLocation  *location,
+	   const FuThunderboltFwObject    *fw,
+	   gboolean                       *val,
+	   GError                        **error)
+{
+	g_autoptr(GByteArray) read = read_location (location, fw, error);
+	if (read == NULL)
+		return;
+	for (gsize i = 0; i < read->len; i++)
+		if (read->data[i] != 0) {
+			*val = TRUE;
+			return;
+		}
+	*val = FALSE;
+	return;
+}
+
+static guint16
+read_uint16 (const FuThunderboltFwLocation  *location,
+	     const FuThunderboltFwObject    *fw,
+	     GError                        **error)
+{
+	guint16 value = 0;
+	g_autoptr(GByteArray) read = read_location (location, fw, error);
+	g_assert_cmpuint (location->len, ==, sizeof (guint16));
+	if (read == NULL)
+		return 0;
+
+	memcpy (&value, read->data, read->len);
+	return GUINT16_FROM_LE (value);
+}
+
+static guint32
+read_uint32 (const FuThunderboltFwLocation  *location,
+	     const FuThunderboltFwObject    *fw,
+	     GError                        **error)
+{
+	guint32 value = 0;
+	g_autoptr(GByteArray) read = read_location (location, fw, error);
+	g_assert_cmpuint (location->len, ==, sizeof (guint32));
+	if (read == NULL)
+		return 0;
+
+	memcpy (&value, read->data, read->len);
+	return GUINT32_FROM_LE (value);
+}
+
+/*
+ * Size of ucode sections is uint16 value saved at the start of the section,
+ * it's in DWORDS (4-bytes) units and it doesn't include itself. We need the
+ * offset to the next section, so we translate it to bytes and add 2 for the
+ * size field itself.
+ *
+ * offset parameter must be relative to digital section
+ */
+static inline guint16
+read_ucode_section_len (guint32 offset, const FuThunderboltFwObject *fw, GError **error)
+{
+	const FuThunderboltFwLocation section_size = { .offset = offset, .len = 2 };
+	guint16 size = read_uint16 (&section_size, fw, error);
+	if (*error != NULL)
+		return 0;
+	return sizeof (guint32) * size + sizeof (size);
+}
+
+/*
+ * Takes a FwObject and fills its section array up
+ * Assumes sections[DIGITAL_SECTION].offset is already set
+ */
+static gboolean
+read_sections (FuThunderboltFwObject *fw, gboolean is_host, guint gen, GError **error)
+{
+	const FuThunderboltFwLocation arc_params_offset = { .offset = 0x75,  .len = 4 };
+	const FuThunderboltFwLocation drom_offset       = { .offset = 0x10E, .len = 4 };
+
+	if (gen >= 3 || gen == 0)
+		fw->sections[DROM_SECTION] = read_uint32 (&drom_offset, fw, error) +
+					     fw->sections[DIGITAL_SECTION];
+	if (*error != NULL)
```

Comment:
This is kinda ugly, I think you have to check the return value of `read_uint32` and return if that's invalid.

Reply:
I can't know if it's valid or not (this was in the commit description of this change :) ).
For `gen < 3`, this location holds a value but it isn't the DROM offset.

Reply:
Oh, sorry, you talked about the `if (*error != NULL)`, not the if (gen >= 3 || gen == 0)`.
So in this case I can assume 0 is invalid, fine.

Reply:
(But then I have to change my precious `minimal-fw` :( )

Reply:
The problem with doing `if (*error != NULL)` is that error can legally be `NULL` and then this is:

a) going to explode with a NULL deref
b) going to "succeed" rather than return FALSE.

Reply:
I understand this, but:
1. This is an internal function and the only caller (plugins/thunderbolt/fu-plugin-thunderbolt.c#L759) passes a valid pointer, and
2. as 0 can be a valid value in some of these cases, the only solution is to make the whole call chain uglier, having boolean return value to signal success/failure and the main return value passed by an out parameter.
For `read_bool ()` I did it, following the documentation guideline:
> By convention, if you return a boolean value indicating success then TRUE means success and FALSE means failure. Avoid creating functions which have a boolean return value and a GError parameter, but where the boolean does something other than signal whether the GError is set. Among other problems, it requires C callers to allocate a temporary error. Instead, provide a "gboolean *" out parameter.

https://developer.gnome.org/glib/stable/glib-Error-Reporting.html

But for the rest, I used the hint from the same page:
> Note that err != NULL in this example is a reliable indicator of whether g_file_get_contents() failed

If you insist, I'll follow your guideline, but I think it worth a second thought.

Reply:
Right, the first quoted section is correct -- it's good GLib practice to return success/failure and return the value byref. There are only a handful of functions in GLib/Gtk that need the GError trick and people pretty much hate those as you *have* to use a going-to-exist local GError and just copy the value on failure.

Reply:
OK. Changing the code.

(I'm a C++ developer and I want my exceptions back. :( )